backend/dataall/modules/dataset_sharing/services/share_managers/s3_bucket_share_manager.py

```diff
@@ -0,0 +1,443 @@
+        return bucket_policy
+
+    def get_bucket_owner_roleid(self):
+        exceptions_roleId = [f'{item}:*' for item in SessionHelper.get_role_ids(
+            self.source_account_id,
+            [self.dataset_admin, self.source_env_admin, SessionHelper.get_delegation_role_arn(self.source_account_id)]
+        )]
+        return exceptions_roleId
+
+    def grant_role_bucket_policy(self):
+        """
+        This function will update bucket policy by granting admin access to dataset admin, pivot role
+        and environment admin along with read only access to accepted share roles. All the policies will only be added
+        once.
+        :return:
+        """
+        logger.info(
+            f'Granting access via Bucket policy for {self.bucket_name}'
+        )
+        try:
+            target_requester_arn = self.get_role_arn(self.target_account_id, self.target_requester_IAMRoleName)
+            bucket_policy = self.get_bucket_policy_or_default()
+            counter = count()
+            statements = {item.get("Sid", next(counter)): item for item in bucket_policy.get("Statement", {})}
+            if DATAALL_READ_ONLY_SID in statements.keys():
+                logger.info(f'Bucket policy contains share statement {DATAALL_READ_ONLY_SID}, updating the current one')
+                statements[DATAALL_READ_ONLY_SID] = self.add_target_arn_to_statement_principal(statements[DATAALL_READ_ONLY_SID], target_requester_arn)
+            else:
+                logger.info(f'Bucket policy does not contain share statement {DATAALL_READ_ONLY_SID}, generating a new one')
+                statements[DATAALL_READ_ONLY_SID] = self.generate_default_bucket_read_policy_statement(self.bucket_name, target_requester_arn)
+
+            if DATAALL_ALLOW_OWNER_SID not in statements.keys():
+                statements[DATAALL_ALLOW_OWNER_SID] = self.generate_owner_access_statement(self.bucket_name, self.get_bucket_owner_roleid())
+
+            bucket_policy["Statement"] = list(statements.values())
+            s3_client = S3Client(self.source_account_id, self.source_environment.region)
+            s3_client.create_bucket_policy(self.bucket_name, json.dumps(bucket_policy))
+        except Exception as e:
+            logger.exception(
+                f'Failed during bucket policy management {e}'
+            )
+            raise e
+
+    def add_target_arn_to_statement_principal(self, statement, target_requester_arn):
+        principal_list = self.get_principal_list(statement)
+        if f"{target_requester_arn}" not in principal_list:
+            principal_list.append(f"{target_requester_arn}")
+        statement["Principal"]["AWS"] = principal_list
+        return statement
+
+    @staticmethod
+    def generate_owner_access_statement(s3_bucket_name, owner_roleId):
+        owner_policy_statement = {
+            "Sid": "AllowAllToAdmin",
```

Comment:
Let's use `DATAALL_ALLOW_OWNER_SID` rather than hardcoded `AllowAllToAdmin` here?

Reply:
And also maybe pass in for `generate_default_bucket_policy()` call

Reply:
Used DATAALL_ALLOW_OWNER_SID rather than hardcoded AllowAllToAdmin. Not sure what you mean by 'pass in for generate_default_bucket_policy() call' 
Its defined as a constant at the top of the file, so we dont need to pass it in.

Reply:
If the bucket policy does not exist we call `generate_default_bucket_policy()` 

the `def generate_default_bucket_policy(s3_bucket_name, owner_role_Id)` has a piece where it adds a statement with `"Sid": "AllowAllToAdmin",`

Could we define the function as:

`def generate_default_bucket_policy(s3_bucket_name, owner_role_Id, allow_owner_sid)` 

...and when calling the function in bucket_share_manager have it as:

`S3ControlClient.generate_default_bucket_policy(self.bucket_name, exceptions_roleId, DATAALL_ALLOW_OWNER_SID)`